### ISSUE TITLE
DRY up the codebase

### ISSUE THREAD

**Issue by danthareja** [danthareja/contribute-to-open-source#1]
> ### Welcome new contributors!
>
> In this experience, you are a **doer**. You take action first, and learn about the action afterwards. This design helps you get comfortable in the unknown, an entity that is omnipresent in the coding world. If you're feeling confused, that's ok. Read the instructions carefully, trust the process, and have fun!
>
> To help guide you through the experience, there will be comment blocks (like the one your reading now) that explains *just enough* context before your next action.
>
> ### Open Source and GitHub
> An **open source** project is any code that can be freely used, changed, and shared by anyone. An open source project usually contains a [license](https://choosealicense.com/) that declares specific terms of use. Open source projects are important because they help develop and spread great ideas and push technology forward.
>
> **GitHub** is a website that hosts open source projects. A project on GitHub is called a **repository** (or **repo**), a folder that holds code. GitHub allows people to collaborate on a repo through features like **issues** and **pull requests**. By the end of this experience, you'll understand how to use these features to contribute your ideas to other projects.
>
> ### Issues
> **Issues** (like this one) allow people to track and discuss any task related to the project. Issues are not always problems with the code. They can also be feature requests or questions about how to use the code.
>
> To help organize the project, issues can be tagged with one or more **label**. When you graduate from this project and are looking for more ways to contribute, look for projects with issues tagged with labels `help wanted`, `good first issue`, `up-for-grabs`, `first-timers-only`, or something similarly inviting.
>
> ### Your turn
> Now that you have a bit of context about the experience, continue reading the body of the issue to learn how to fix it.
---

We're off to an amazing start disrupting the calculator industry. Our beautiful contributors burned all of last week's midnight oil to release our first version and the initial reception has been unprecedented!

After a code review from some experts, we've learned that there's some unnecessary repetition in the codebase. Removing this repetition will make our codebase easier to maintain as we continue to grow into the world's best calculator company.

The team is burnt out from last week's push and needs your help! Before blowing half of our capital investment on a vacation in the Swiss Alps, the team managed to write some tests for this refactor, and left behind a roadmap to the next big release:

- [ ] Read the [contributing guidelines](https://github.com/danthareja/contribute-to-open-source/blob/master/CONTRIBUTING.md)
- [ ] [Build the project locally](https://github.com/danthareja/contribute-to-open-source/blob/master/CONTRIBUTING.md#build-the-project-locally)
- [ ] Go to [this line](https://github.com/danthareja/contribute-to-open-source/blob/master/src/calculator.test.js#L4) and change `describe.skip` to `describe.only`
- [ ] Run `npm test`
- [ ] Notice that the tests are failing
- [ ] Implement [`_check`](https://github.com/danthareja/contribute-to-open-source/blob/master/src/calculator.js#L1) to make the tests pass
- [ ] Change the `describe.only` to `describe` so all the tests run
- [ ] Run `npm test` again
- [ ] Make sure that all the tests are passing
- [ ] [Submit a pull request](https://github.com/danthareja/contribute-to-open-source/blob/master/CONTRIBUTING.md#submit-a-pull-request)

> ---
> This experience can be completed my multiple people. Below, you'll see [references](https://help.github.com/articles/autolinked-references-and-urls/#issues-and-pull-requests) to this issue from various other contributors. You can ignore them.

**Comment by sourabh1031** [comment]
i am working on it and facing eslint problems since i am working on windows do i need to take care special care?

**Comment by danthareja** [comment]
@sourabh1031, are they the same problems you mentioned in your other issue https://github.com/danthareja/contribute-to-open-source-server/issues/2? If so, did you try what I mentioned in the reply?

**Comment by sourabh1031** [comment]
Yes it got solved by your help thanks and now I must switch my work onto
Unix based system isn't?

**Comment by danthareja** [comment]
No need. As long as the styling issues are fixed, you can keep developing on Windows.

**Comment by sourabh1031** [comment]
![2018-01-14](https://user-images.githubusercontent.com/25590627/34913349-4a21726a-f920-11e7-8e36-0e69224b5971.png)
Please see this i am not able to proceed after it

**Comment by danthareja** [comment]
Hey @Ri2parna, I appreciate you asking if this is issue is still open. This is a very good practice when contributing to projects. 

This project is slightly different than normal in that it's always open for contribution for everyone. Instead of contributing your work back to the master branch, you'll contribute to the branch that i've made for you.

Follow the instructions above and have at it!

**Comment by scelestus1** [comment]
Hey @danthareja, just wanted to say thanks for putting this little practice scenario together. 

I've wanted to get into open source contribution for a while now, but was too intimidated by the Git arena to even get started. However, after going through your training module, I now feel like I have a better understanding of what's going on/how things are done and am confident enough to try to help out on some "real" projects!

**Comment by danthareja** [comment]
I'm glad you enjoyed it @scelestus1, Have fun contributing out there!

**Comment by MisterAwesome23** [comment]
Aaaaaaaa! I did it. My first PR accepted and the message saying "You did it. You are an open source developer now" is AN ABSOLUTE BLISS! 
Thank you so much for such amazing and well-detailed guidelines. Means a lot.
Thank you.

**Comment by MisterAwesome23** [comment]
Just one question, while contributing to real projects the only difference is in the last step that is creating the pull request across the forks, instead of changing the first base from master to ourselves(our username) we should keep that to master right? Correct me if I am wrong.

**Comment by niteshnanda02** [comment]
![screenshot 28](https://user-images.githubusercontent.com/40588354/50728421-72f10f00-114f-11e9-901e-6dd1477efe70.png)
In node.js the project is not install.What to do?

**Comment by danthareja** [comment]
@niteshnanda02, did you figure this one out? Based on the screenshot, it looks like you are inside of the node command-line. If you follow the instruction to exit the (press Control+D), you should exit the node command-line, back to your regular terminal window. Once in the regular terminal window, you should be able to type `npm install` again.

**Comment by danthareja** [comment]
@MisterAwesome23, apologies for the delay in getting back to you.

> Aaaaaaaa! I did it. My first PR accepted and the message saying "You did it. You are an open source developer now" is AN ABSOLUTE BLISS!
> Thank you so much for such amazing and well-detailed guidelines. Means a lot.
> Thank you.

This makes me smile so much, I'm glad you enjoyed the process!

> Just one question, while contributing to real projects the only difference is in the last step that is creating the pull request across the forks, instead of changing the first base from master to ourselves(our username) we should keep that to master right? Correct me if I am wrong.

Yes! By convention, most projects live on the `master` branch, so you would be correct in this case. This is not always the case though! Do make sure to reference the individual project's contributing guidelines to see if they require any special steps.

**Comment by danthareja** [comment]
> Just one question, while contributing to real projects the only difference is in the last step that is creating the pull request across the forks, instead of changing the first base from master to ourselves(our username) we should keep that to master right? Correct me if I am wrong.

**Comment by niteshnanda02** [comment]
What's the error in this?
![screenshot 29](https://user-images.githubusercontent.com/40588354/51077600-ee623b80-16ce-11e9-9e50-ece5509a2507.png)

**Comment by danthareja** [comment]
@niteshnanda02, this is the expected error message you will receive the first time you run `npm test`. Your job now is to make the red tests turn green by writing code in `calculator.js` that satisfies the test cases. If you're confused on exactly what the tests are doing, check out the documentation for the test libraries: [Mocha](https://mochajs.org/) and [Chai](https://www.chaijs.com/) and [Sinon](https://sinonjs.org/)

**Comment by sog01** [comment]
Yeaaayy, my PR already merged. It's very excited even has been merged by bot.

**Comment by techmariah** [comment]
@danthareja Did I make the pull request in the wrong location since it is now closed by you? If so, should I apologies. I will pull the request in my forked repo. Thank you.

**Comment by techmariah** [comment]
@danthareja Did I make the pull request in the wrong location since it is now closed by you?

**Comment by vinayaksharma082** [comment]
Yess, my first commit , 
i am feeling so great !

**Comment by Xyno18** [comment]
I'm not familiar with JavaScript at all. You're asking me to implement _check(), but how to do that?

**Comment by danthareja** [comment]
Good question, @Xyno18!

There are a few comments that describe how to implement `_check` inside of the file [`src/calculator.js`](https://github.com/danthareja/contribute-to-open-source/blob/master/src/calculator.js#L2). If these comments are confusing, I'd recommend starting with a free [introduction to JavaScript](https://www.codecademy.com/learn/introduction-to-javascript) course to learn the basic syntax of the language. Once you're familiar with variables, conditional statements, and functions, you should come back here.

**Comment by richard937** [comment]
![whyerror](https://user-images.githubusercontent.com/42942897/62817313-f34a7900-bb51-11e9-9eae-cb03e2d4d26f.png)

why this error?

**Comment by qaws01395** [comment]
Hey,
Question: On this page [Open a Pull Request](https://github.com/danthareja/contribute-to-open-source/blob/master/CONTRIBUTING.md#open-a-pull-request) part,

* This is a weird requirement. In the real-world, most projects will expect you to open a Pull Request against the master base branch. Can you guess why it's needed here?


Why would be the reason?

**Comment by danthareja** [comment]
Good question @qaws01395!

This requirement is necessary here in order to allow multiple people to contribute the same code. Without this requirement, once one contributor's PR was merged into master, no one else would be able to merge the same code.

**Comment by qaws01395** [comment]
Thanks for the explanation!

**Comment by jpkim921** [comment]
Hi,

I'm new to open source and would like to contribute.

**Comment by danthareja** [comment]
Welcome to open-source, @jpkim921!

Have you completed the exercise from this repo? If not, that's a great place to start -- just complete the instructions in the issue's body above. Once you complete the exercise, you'll get an automated comment in your pull request with some ideas for your next contribution.

If you have further questions, feel free to ask here.

**Comment by HakaCode** [comment]
I'm new to Open Source collaboration, I can contribute?

**Comment by Om-Vg** [comment]
New to OpenSource and would like to contribute

**Comment by danthareja** [comment]
Yes! @HakaCode and @OmLakshmi - you both are very welcome to contribute.

Please head over to issue #1 and follow the instructions there.

**Comment by ghost** [comment]
OKay, let me have a gpo @danthareja

**Comment by ghost** [comment]
Is this issue still active @danthareja

**Comment by danthareja** [comment]
Yes! Everyone can contribute their own solution to this issue @KarenEfereyan. Make sure you read the instructions about submitting a PR to your own branch, and have fun!

**Comment by rondog64** [comment]
Thank you ❤️ great job

**Comment by rondog64** [comment]
Amazing

**Comment by rondog64** [comment]
Team work

**Comment by syspentarc** [comment]
I've got to be honest, it has been so long since I worked with NodeJS that I feel kind of afraid of messing everything up including my Linux install. Called it... 
![2020-12-28-04:31:22-screenshot](https://user-images.githubusercontent.com/76585333/103189662-accff880-48c5-11eb-92d6-2740d07a0955.png)
@danthareja I'm having some problems with the dependencies.

**Comment by ghost** [comment]
I had some problems with the dependencies, but I think I managed to fix it. Still, kind of afraid of messing everything up.

**Comment by ghost** [comment]
I haven't had this much fun debugging in a long time! I'm not really sure if I've ever had this much fun debugging, to be honest. Haven't fixed it yet, but really loving it. Thank you!

**Comment by ghost** [comment]
I made the PR, thanks for making this. Loved working on it!

**Comment by danthareja** [comment]
Nice work @syspentarc, thanks for playing!

**Comment by ghost** [comment]
I'm new to GitHub, but I've done one or two PRs before and did my research prior to joining the platform. This really gave me more confidence in working on doing more of these types of contributions instead of looking more to documentation to contribute to.

**Comment by ghost** [comment]
Please consider doing more like this, I'd love to see more stuff exactly like this. :smile:

**Comment by Satsuki39** [comment]
it's unfair.and I don't run in your race.

**Comment by satcasm** [comment]
Hi, when I am running "git push origin master" it is showing :
remote: Permission to danthareja/contribute-to-open-source.git denied to satcasm.
fatal: unable to access 'https://github.com/danthareja/contribute-to-open-source.git/': The requested URL returned error: 403
I tried to create a local branch and then push it but I am still getting the same error.

**Comment by satcasm** [comment]
Hi, I am new to contributing to open source and it was a fun little assignment. I just wanted clarification on one thing, we are selecting the base branch with our name so that we don't make changes to the  master branch of the repo . Is that so?

**Comment by danthareja** [comment]
> Hi, I am new to contributing to open source and it was a fun little assignment. I just wanted clarification on one thing, we are selecting the base branch with our name so that we don't make changes to the master branch of the repo . Is that so?

Yeah, that's exactly right. That's a distinct difference between this project and one you'll find out in the real world.

If you made changes to the master branch of the repo, only one person would be able to contribute, or we'd have to do some complex revert/pulling which could be hard to understand for new contributors.

**Comment by larabee7777** [comment]
I would like to contribute if possible?

**Comment by TB9652** [comment]
I get a 185 errors when I run "npm run lint".  The errors are all linebreak errors as follows     "error  Expected linebreaks to be 'LF' but found 'CRLF'  linebreak-style"
![Error](https://user-images.githubusercontent.com/70457523/115933950-d8628a00-a444-11eb-98bc-3e8e1fe446b0.PNG)

**Comment by chincholikarsalil** [comment]
Thanks for helping us beginners to take our first steps. 
Appreciate your efforts 😄

**Comment by TB9652** [comment]
Sorry for the stupid question, but how do I take care of the errors?  Have been following the instructions on this but I'm a bit lost now so any guidance would be appreciated.

**Comment by chincholikarsalil** [comment]
> Sorry for the stupid question, but how do I take care of the errors? Have been following the instructions on this but I'm a bit lost now so any guidance would be appreciated.

@TB9652 
According to the image that you have posted, I think that you are using Windows to build the project. I recommend using Linux based OS like Ubuntu. 
Also, if you do not have Ubuntu, you can always try using the Windows Subsystem for Linux (WSL), it is very convenient.

**Comment by danthareja** [comment]
hey @TB9652, sorry for the delay here. there are no stupid questions!

windows and mac have two different default line endings. i authored this on a mac, which defaults to the 'CLRF' line ending. since you're on windows, you're seeing the error message, which is asking you to write to the mac line endings.

to fix the errors, try running `npm run lint -- --fix`, which tells eslint to self-correct these endings.

**Comment by chincholikarsalil** [comment]
> to fix the errors, try running `npm run lint -- --fix`, which tells eslint to self-correct these endings.

@danthareja @TB9652 
Sorry, had no idea about this command.

**Comment by TB9652** [comment]
Thanks for the guidance.  I will try that and let you know how it goes.  I'm not very well schooled on using Git Bash.  I have more experience using Microsoft VS so this is a true learning experience for me.

**Comment by mspaansen** [comment]
Ok

**Comment by Kamalesh3112** [comment]
Committed to the issue and contributed

**Comment by Kodeiak** [comment]
@danthareja , I get "command not found" when trying to run "npm install" in git.  I have downloaded node.js and am in my clone of contribute-to-open-source

**Comment by darkranger778** [comment]
> > ### Welcome new contributors!
> > In this experience, you are a **doer**. You take action first, and learn about the action afterwards. This design helps you get comfortable in the unknown, an entity that is omnipresent in the coding world. If you're feeling confused, that's ok. Read the instructions carefully, trust the process, and have fun!
> > To help guide you through the experience, there will be comment blocks (like the one your reading now) that explains _just enough_ context before your next action.
> > ### Open Source and GitHub
> > An **open source** project is any code that can be freely used, changed, and shared by anyone. An open source project usually contains a [license](https://choosealicense.com/) that declares specific terms of use. Open source projects are important because they help develop and spread great ideas and push technology forward.
> > **GitHub** is a website that hosts open source projects. A project on GitHub is called a **repository** (or **repo**), a folder that holds code. GitHub allows people to collaborate on a repo through features like **issues** and **pull requests**. By the end of this experience, you'll understand how to use these features to contribute your ideas to other projects.
> > ### Issues
> > **Issues** (like this one) allow people to track and discuss any task related to the project. Issues are not always problems with the code. They can also be feature requests or questions about how to use the code.
> > To help organize the project, issues can be tagged with one or more **label**. When you graduate from this project and are looking for more ways to contribute, look for projects with issues tagged with labels `help wanted`, `good first issue`, `up-for-grabs`, `first-timers-only`, or something similarly inviting.
> > ### Your turn
> > Now that you have a bit of context about the experience, continue reading the body of the issue to learn how to fix it.
> 
> We're off to an amazing start disrupting the calculator industry. Our beautiful contributors burned all of last week's midnight oil to release our first version and the initial reception has been unprecedented!
> 
> After a code review from some experts, we've learned that there's some unnecessary repetition in the codebase. Removing this repetition will make our codebase easier to maintain as we continue to grow into the world's best calculator company.
> 
> The team is burnt out from last week's push and needs your help! Before blowing half of our capital investment on a vacation in the Swiss Alps, the team managed to write some tests for this refactor, and left behind a roadmap to the next big release:
> 
> * [ ]  Read the [contributing guidelines](https://github.com/danthareja/contribute-to-open-source/blob/master/CONTRIBUTING.md)
> * [ ]  [Build the project locally](https://github.com/danthareja/contribute-to-open-source/blob/master/CONTRIBUTING.md#build-the-project-locally)
> * [ ]  Go to [this line](https://github.com/danthareja/contribute-to-open-source/blob/master/src/calculator.test.js#L4) and change `describe.skip` to `describe.only`
> * [ ]  Run `npm test`
> * [ ]  Notice that the tests are failing
> * [ ]  Implement [`_check`](https://github.com/danthareja/contribute-to-open-source/blob/master/src/calculator.js#L1) to make the tests pass
> * [ ]  Change the `describe.only` to `describe` so all the tests run
> * [ ]  Run `npm test` again
> * [ ]  Make sure that all the tests are passing
> * [ ]  [Submit a pull request](https://github.com/danthareja/contribute-to-open-source/blob/master/CONTRIBUTING.md#submit-a-pull-request)
> 
> > This experience can be completed my multiple people. Below, you'll see [references](https://help.github.com/articles/autolinked-references-and-urls/#issues-and-pull-requests) to this issue from various other contributors. You can ignore them.

Implement check #27

**Comment by raynardmiot** [comment]
How do you get your username to show up in the base ref when opening a new pull request?

**Comment by prasannajeet2002** [comment]
#1512 how to implement check

**Comment by qais-aziz** [comment]
@danthareja why is my pull request closed?

**Comment by danthareja** [comment]
> @danthareja why is my pull request closed?

As mentioned in the [contributing guidelines](https://github.com/danthareja/contribute-to-open-source/blob/master/CONTRIBUTING.md#open-a-pull-request), I can only accept a pull request made to _your branch_.

This is a weird requirement, but it's necessary here in order to allow multiple people to contribute the same code. In the real world, most projects would expect you to open a Pull Request against the `master` base branch.

### [**Click here to open this pull request to your branch**](https://github.com/danthareja/contribute-to-open-source/compare/qais-aziz...qais-aziz:qais-contributes)

**Comment by aakansharamesh42** [comment]
Hi, in the npm run lint I am getting the following error:

![img](https://user-images.githubusercontent.com/68271524/188409127-8f7b9f8a-cdce-40a6-ac60-2e9ccd90db64.png)

**Comment by marcostech** [comment]
> @sourabh1031, are they the same problems you mentioned in your other issue [danthareja/contribute-to-open-source-server#2](https://github.com/danthareja/contribute-to-open-source-server/issues/2)? If so, did you try what I mentioned in the reply?

I believe this should be on the Contributors.md since we, begginners, dont understand this erros yet and ive run into this "problem" myself, or maybe we should at least be directed on how to solve then.

**Comment by sh471** [comment]
@danthareja

**Comment by alok-38** [comment]
Could you please assign this task to me?

**Comment by sh471** [comment]
sure

*Best Regards,*

*Sherif Hikal*

**Comment by amansingh179** [comment]
Can i work on this project?

**Comment by renovX** [comment]
this still an issue ?

**Comment by anubhav1206** [comment]
Can i work on this?

**Comment by sh471** [comment]
Hello 👋

**Comment by ghost** [comment]
@danthareja There is a typo in the issue description

> This experience can be completed my multiple people.

This experience can be completed by multiple people. 

![typo-in-issue-description](https://github.com/danthareja/contribute-to-open-source/assets/136180642/cf66e64a-c394-4209-871f-e4f313c05dfc)

**Comment by danthareja** [comment]
> @danthareja There is a typo in the issue description

Great catch! Thanks for the contribution :)

**Comment by bakhtiar56** [comment]
It is said to implement __checks to make tests pass but when clicking on that link, I am directed to 
exports._check = () => {
  // DRY up the codebase with this function
  // First, move the duplicate error checking code here
  // Then, invoke this function inside each of the others
  // HINT: you can invoke this function with exports._check()
};

What do these instructions mean? I am unclear about them as to what do I have to write in the codeblock

**Comment by barazageorge** [comment]
Just ran and re-ran the tests as per the instructions and noticed there was a change in the package-lock.json file. Did nothing on the code base though.

**Comment by fury2587** [comment]
Hello I am a newcomer, i  have make a pull request as mentioned in the guideline but i don't know that you are able to see or not.       please help me in that.
![Uploading Screenshot (18).png…]()

**Comment by fury2587** [comment]
![Screenshot (18)](https://github.com/user-attachments/assets/90831e1b-3157-481a-8d18-51011e2a13d2)

**Comment by JashanpreetSingh25** [comment]
For anyone having a " Expected linebreaks to be 'LF' but found 'CRLF' " error, this is just a related to inconsistent line endings in the code. It is just a Windows problem because Windows uses CRLF (Carriage Return + Line Feed) while Unix based system like Linux and MacOS uses LF (Line Feed). For Windows users, you can use WSL or use article to know how to fix it https://medium.com/bootdotdev/how-to-get-consistent-line-breaks-in-vs-code-lf-vs-crlf-e1583bf0f0b6. This was scary as a beginner but I just googled it, and done.

**Comment by Mariam-Saeed** [comment]
![image](https://github.com/user-attachments/assets/9579649f-a379-4392-8e34-3f69e05ffd48)
How can I fix these errors?

**Comment by EvanPrograms** [comment]
There is one issue with the readme I think. You dont talk about creating your own branch? That could be added?

**Comment by ddlite92** [comment]
** EDITED ** 
- I just committed and pushed the required files, which are calculator and calculator.test, and everything is fine.
- If anyone else also realizes about the package-lock.json, I think it is safe for us to ignore at the moment since we are beginners anyway. 

Hi, want to ask.
I have 3 files not staged while the documentation only said 2 which is the calculator & calculator.test
I don't think I edited any of json file. Should I staged all or just the 2 js file?

![Image](https://github.com/user-attachments/assets/dc25b196-9bd5-4b21-adea-d48d4da9648d)

**Comment by SAIRAMSSSS** [comment]
I would like to contribute in this project by solving your issues

**Comment by Gevans4352** [comment]
How do i input here?

**Comment by anilu2660** [comment]
I want to contribute in this project

**Comment by Pantkartik** [comment]
hey! I would like to contribute  in this project
